### PR TITLE
refactor(telemetry-context): Provide internal used telemetry context interface

### PR DESCRIPTION
**Pull request by chentong7**
## Description

> Deprecating `get` and `serialize` on the interface, it should just be a write-only interface for instrumentation (like logger.sendEvent).  Then containerRuntime has the concrete implementation (`TelemetryContext`) which has `serialize` (and perhaps no one would ever really need `get`)

> Providing internal used telemetry context interface

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -3217,14 +3217,14 @@ export class ContainerRuntime
 			if (runGC) {
 				await this.collectGarbage(
 					{ logger: summaryLogger, runSweep, fullGC },
-					telemetryContext,
+					telemetryContext as ITelemetryContext,
```
Comment (alexvy86) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3220]: @markfields @chentong7 what do you think about adding `get()` to `ITelemetryContextExt` but removing it from `ITelemetryContext` (we already removed the other deprecated one, `serialize()`, so why not go all the way to clean up the alpha interface), making the `TelemetryContext` class implement _both_ interfaces (I think we could even use the same set/setMultiple methods to cover both signatures since the non-ext property types are subsets of the new ones, so receiving one of those for `set`-ing should always work), and inside [`incrementTelemetryMetric()`](https://github.com/microsoft/FluidFramework/blob/fcfbf7707d5407927533f9b4da1364c03d0125a6/packages/dds/shared-object-base/src/sharedObject.ts#L763) (the only place that uses `ITelemetryContext.get()`), we confirm that the `get` property exists on the passed-in `telemetryContext`, cast it to `ITelemetryContextExt`, and use it?

Comment (alexvy86) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3220]: In fact, not even have `get()` in the Ext interface but only on the `TelemetryContext` class, and `incrementTelemetryMetric` would cast to the class after checking that `get` exists.

Comment (chentong7) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3220]: I think the idea was keep the public interface only have the set function and let implementation type to customize the get function. I will say keep `ITelemetryContext` only have set(), and move get() and serialize() from `TelemetryContext` to `ITelemetryContextExt`

Comment (alexvy86) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3220]: My point is that we don't need `get()` or `serialize()` in _either_ of the interfaces. I assumed right now we didn't remove it from `ITelemetryContext` because there is a code path that receives that interface and needs to call `get` on it. We definitely still need `get()` on the _class_ to support that code path, but if we remove it from both interfaces, then I think we can avoid the need for casting here, and we can just cast in a much safer way only inside that single function that needs `get()`. It will still receive an object typed as `ITelemetryContext`, but it can check if the object has a `get` property, and if so cast it to the _class_ `TelemetryContext` and use the function. That would keep both interfaces "clean" in the sense that they would only have set/setMultiple.

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3220]: Yeah I agree with what Alex is outlining here

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -3217,14 +3217,14 @@ export class ContainerRuntime
...
 				);
 			}
 
 			const { stats, summary } = await this.summarizerNode.summarize(
 				fullTree,
 				trackState,
-				telemetryContext,
+				telemetryContext as ITelemetryContext,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3227]: Why is this (and the other) cast necessary?

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 3]
```diff
@@ -320,6 +321,33 @@ export interface ISummarizerNodeWithGC extends ISummarizerNode {
  */
 export const channelsTreeName = ".channels";
 
+/**
+ * Contains telemetry data relevant to summarization workflows.
+ * This object is expected to be modified directly by various summarize methods.
```
Comment (markfields) [inline packages/runtime/runtime-definitions/src/summary.ts:326]: Can you add a sentence that contrasts this v. `ITelemetryContext`?

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 4]
```diff
@@ -352,19 +380,9 @@ export interface ITelemetryContext {
 	 * @deprecated This interface should only be used for instrumenting, not for attempting to read already-set telemetry data.
 	 *
 	 * @param prefix - unique prefix for this data (ex: "fluid:map:")
-	 * @param property - property name of the telemetry data being tracked (ex: "DirectoryCount")
 	 * @returns undefined if item not found
 	 */
 	get(prefix: string, property: string): TelemetryBaseEventPropertyType;
```
Comment (markfields) [inline packages/runtime/runtime-definitions/src/summary.ts:385]: As discussed in the other thread with Alex, this should be removed here.  Anyone internal who needs it should cast to the class.

That said, we can also learn from cases like the one that is using this, to think of other functionality the general-purpose public API would need.  e.g. we could consider adding an `increment` function or something at some point.

File: pnpm-lock.yaml [hunk 5]
```diff
@@ -18995,7 +18997,7 @@ packages:
     dependencies:
       '@inquirer/type': 1.2.1
       '@types/mute-stream': 0.0.4
-      '@types/node': 20.12.6
+      '@types/node': 20.12.10
```
Comment (markfields) [inline pnpm-lock.yaml:19000]: Probably revert by hand these changes to this file that are not about adding the telemetry-utils dependency.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (chentong7) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3220]: Gotcha. Updating accordingly

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (chentong7) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: No longer needed after remove get()

File: packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md [hunk 6]
```diff
@@ -417,14 +418,16 @@ export interface ISummaryTreeWithStats {
 
 // @public
 export interface ITelemetryContext {
-    // @deprecated
-    get(prefix: string, property: string): TelemetryBaseEventPropertyType;
-    // @deprecated
-    serialize(): string;
```
Comment (alexvy86) [inline packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md:423]: Let's add a changeset announcing the removal of these properties since it's a breaking change.

File: packages/service-clients/azure-client/api-report/azure-client.api.md [hunk 7]
```diff
@@ -67,7 +67,7 @@ export interface AzureContainerVersion {
 
 // @internal @deprecated
 export class AzureFunctionTokenProvider implements ITokenProvider {
-    constructor(azFunctionUrl: string, user?: Pick<AzureMember<any>, "name" | "id" | "additionalDetails"> | undefined);
+    constructor(azFunctionUrl: string, user?: Pick<AzureMember<any>, "name" | "additionalDetails" | "id"> | undefined);
```
Comment (alexvy86) [inline packages/service-clients/azure-client/api-report/azure-client.api.md:70]: Revert this one? One of those weird build issues that we see but aren't related to this PR.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 8]
```diff
@@ -777,8 +778,10 @@ export abstract class SharedObject<
 		incrementBy: number,
 		telemetryContext?: ITelemetryContext,
 	) {
-		const prevTotal = (telemetryContext?.get(this.telemetryContextPrefix, propertyName) ??
-			0) as number;
+		const prevTotal = ((telemetryContext as TelemetryContext)?.get(
```
Comment (alexvy86) [inline packages/dds/shared-object-base/src/sharedObject.ts:781]: We can be more type safe here. I think this should work, can you verify? Will need to import our `assert` function.
```suggestion
		// TelemetryContext is the only current (and planned) implementation of ITelemetryContext
		assert(telemetryContext instanceof TelemetryContext, "received context is not an instance of TelemetryContext");
		const prevTotal = (telemetryContext.get(
```

**Review by alexvy86 — COMMENTED** [review]
A couple of final things but this is looking almost done!

File: packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md (hunk 6 above)
Comment (chentong7) [reply packages/runtime/runtime-definitions/api-report/runtime-definitions.api.md:423]: On it

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 8 above)
Comment (chentong7) [reply packages/dds/shared-object-base/src/sharedObject.ts:781]: Updated

File: packages/runtime/runtime-utils/src/summaryUtils.ts [hunk 9]
```diff
@@ -412,13 +412,13 @@ export function processAttachMessageGCData(
 /**
  * @internal
  */
-export class TelemetryContext implements ITelemetryContext {
-	private readonly telemetry = new Map<string, TelemetryBaseEventPropertyType>();
+export class TelemetryContext implements ITelemetryContextExt {
```
Comment (alexvy86) [inline packages/runtime/runtime-utils/src/summaryUtils.ts:415]: We should keep implementing both interfaces, so we make sure that the class doesn't accidentally stops being compatible with `ITelemetryContext`, since we're still passing it in code paths that receive `ITelemetryContext`. This should be doable without adding any more methods, if the `value(s)` parameter to `set()` and `setMultiple()` takes `TelemetryEventPropertyTypeExt | TelemetryBaseEventPropertyType`.
```suggestion
export class TelemetryContext implements ITelemetryContext, ITelemetryContextExt {
```

**Review by alexvy86 — COMMENTED** [review]
A few more comments below. I saw that the some of the failing tests in CI are hitting the new assert about the input parameter not being an instance of `TelemetryContext`. Let's debug them to see what exactly are they getting, it might be a mock context, in which case the assert should just check for `'get' in telemetryContext` or something similar that guarantees the function exists without constraining us to explicitly wanting the `TelemetryContext` class.

**Comment by chentong7** [comment]
> A few more comments below. I saw that the some of the failing tests in CI are hitting the new assert about the input parameter not being an instance of `TelemetryContext`. Let's debug them to see what exactly are they getting, it might be a mock context, in which case the assert should just check for `'get' in telemetryContext` or something similar that guarantees the function exists without constraining us to explicitly wanting the `TelemetryContext` class.

Yea. It's thrown on `FluidFramework/packages/dds/shared-object-base/src/sharedObject.ts:782` that asserting "telemetryContext instanceof TelemetryContext". It should be fine after switching TelemetryContext to implement both. Updating...

File: packages/runtime/runtime-definitions/package.json [hunk 10]
```diff
@@ -113,8 +114,7 @@
 				"forwardCompat": false,
 				"backCompat": false
 			},
-			"RemovedInterfaceDeclaration_IFluidDataStoreContextEvents": {
```
Comment (alexvy86) [inline packages/runtime/runtime-definitions/package.json:116]: So, it seems that `IFluidDataStoreContextEvents` had been incorrectly added here, right? It didn't change in the previous PR, so removing it is actually correct. Let's add that to the PR description if so.

File: pnpm-lock.yaml [hunk 11]
```diff
@@ -9005,6 +9006,7 @@ importers:
       '@fluidframework/driver-definitions': link:../../common/driver-definitions
       '@fluidframework/id-compressor': link:../id-compressor
       '@fluidframework/protocol-definitions': 3.2.0
+      '@fluidframework/telemetry-utils': link:../../utils/telemetry-utils
```
Comment (markfields) [inline pnpm-lock.yaml:9009]: You can/should revert the changes to this file besides this and the one above.  It's fine to bump those dependencies, but I'd strongly recommend doing it in a separate PR for clarity.

File: packages/runtime/runtime-definitions/package.json (hunk 10 above)
Comment (markfields) [reply packages/runtime/runtime-definitions/package.json:116]: Seems totally unrelated to this change, if so I'd suggest cleaning this up in a separate PR.  How did this get out of sync anyway?

**Review by markfields — APPROVED** [review]
Looks good!  A couple parts of the diff seem unrelated to the change so I would suggest reverting those first (as noted in comments)

File: packages/runtime/runtime-definitions/package.json (hunk 10 above)
Comment (chentong7) [reply packages/runtime/runtime-definitions/package.json:116]: Restore the file

File: packages/runtime/runtime-definitions/package.json [hunk 12]
```diff
@@ -113,9 +114,11 @@
...
-				"forwardCompat": false,
+			"InterfaceDeclaration_ITelemetryContext": {
 				"backCompat": false
+			},
+			"RemovedInterfaceDeclaration_IFluidDataStoreContextEvents": {
+				"forwardCompat": false
```
Comment (alexvy86) [inline packages/runtime/runtime-definitions/package.json:121]: Here we should still restore the `"backCompat": false` that IFluidDataStoreContextEvents` already had

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 13]
```diff
@@ -343,28 +344,35 @@ export interface ITelemetryContext {
...
 		property: string,
-		values: Record<string, TelemetryBaseEventPropertyType>,
+		values: Record<string, TelemetryEventPropertyTypeExt>,
 	): void;
+}
 
+/**
+ * Contains telemetry data relevant to summarization workflows.
+ * This object is expected to be modified directly by various summarize methods.
+ * @public
```
Comment (alexvy86) [inline packages/runtime/runtime-definitions/src/summary.ts:354]: This used to be alpha, we don't need to make it public, right?

Comment (chentong7) [reply packages/runtime/runtime-definitions/src/summary.ts]: Right right. Modified that by mistake. Thanks for correcting it!

File: packages/runtime/runtime-definitions/package.json (hunk 12 above)
Comment (chentong7) [reply packages/runtime/runtime-definitions/package.json]: Another mistake. Thx!!

**Comment by chentong7** [comment]
> A few more comments below. I saw that the some of the failing tests in CI are hitting the new assert about the input parameter not being an instance of `TelemetryContext`. Let's debug them to see what exactly are they getting, it might be a mock context, in which case the assert should just check for `'get' in telemetryContext` or something similar that guarantees the function exists without constraining us to explicitly wanting the `TelemetryContext` class.

Tested locally. Seems like a lot of our mock runtime use `undefined` TelemetryContext. updating assert accordingly

**Review by alexvy86 — APPROVED** [review]
